esi_leap/objects/owner_change.py

```diff
@@ -0,0 +1,185 @@
+            return cls._from_db_object(context, cls(), db_owner_change)
+
+    @classmethod
+    def get_all(cls, filters, context=None):
+        db_owner_changes = cls.dbapi.owner_change_get_all(filters)
+        return cls._from_db_object_list(context, db_owner_changes)
+
+    def create(self, context=None):
+        updates = self.obj_get_changes()
+
+        @utils.synchronized(
+            utils.get_resource_lock_name(
+                updates['resource_type'], updates['resource_uuid']),
+            external=True)
+        def _create_owner_change():
+
+            if updates['start_time'] >= updates['end_time']:
+                raise exception.InvalidTimeRange(
+                    resource='owner_change',
+                    start_time=str(updates['start_time']),
+                    end_time=str(updates['end_time'])
+                )
+
+            ro = ro_factory.ResourceObjectFactory.get_resource_object(
+                updates['resource_type'], updates['resource_uuid'])
+            ro.verify_availability(updates['start_time'],
+                                   updates['end_time'],
+                                   is_owner_change=True)
+
+            db_owner_change = self.dbapi.owner_change_create(updates)
+            self._from_db_object(context, self, db_owner_change)
+
+        _create_owner_change()
+
+    def fulfill(self, context=None):
+        @utils.synchronized(utils.get_resource_lock_name(self.resource_type,
+                                                         self.resource_uuid),
+                            external=True)
+        def _fulfill_owner_change():
+            LOG.info("Setting owner of %s %s to %s",
+                     self.resource_type, self.resource_uuid, self.to_owner_id)
+
+            resource = self.resource_object()
+            resource.set_owner(self.to_owner_id)
+
+            # activate lease
+            self.status = statuses.ACTIVE
+            self.fulfill_time = datetime.datetime.now()
+            self.save(context)
+
+        _fulfill_owner_change()
+
+    def cancel(self):
+        offers_to_cancel = self.offers()
+        for offer in offers_to_cancel:
+            offer.cancel()
+        leases_to_cancel = self.leases()
+        for lease in leases_to_cancel:
+            lease.cancel()
+
+        @utils.synchronized(utils.get_resource_lock_name(self.resource_type,
+                                                         self.resource_uuid))
+        def _cancel_owner_change():
+            LOG.info("Setting owner of %s %s back to %s",
+                     self.resource_type, self.resource_uuid,
+                     self.from_owner_id)
+
+            resource = self.resource_object()
+            resource.set_owner(self.from_owner_id)
+
+            self.status = statuses.CANCELLED
+            self.expire_time = datetime.datetime.now()
+            self.save(None)
+
+        _cancel_owner_change()
+
+    def expire(self, context=None):
```

Comment:
I guess in theory, for a resource, no offer or lease will be created during an owner change?

Reply:
I have this confusion because creating an offer or a lease calls verify_availability() in resource_objects, and this function checks if there're time conflicts with other offers, leases, and owner changes on this resource and raises exceptions when conflicts happen. Not sure if I understand it correctly.

Reply:
Ah, yes! If you look at `resource_verify_availability` in `db/sqlalchemy/api.py`, it takes a flag: `is_owner_change`. Leases and offers call verify_availability with that flag set to false, meaning that they specifically do *not* check against `owner_changes`; instead they would have used `check_resource_admin` earlier to verify that the lease/offer created is compatible with whoever the admin is at the time, as per the owner change.

Reply:
oh, I see! It makes sense to me now.